### ISSUE TITLE
optimisticResponse reverts data after request failure

### ISSUE THREAD

**Issue by MitchellMonaghan** [aerogear/offix#447]
This might just be a question and not a bug, but it might also be a discussion or a bug.

Basically I have data that gets modified offline. I use offix to store these local offline changes. We push data to the api, the request fails for w/e reason, because this is using apollo optimisticResponses apollo reverts this data in the ui. The user refreshes the page and offix restores the data as it should be. Is there a way i can prevent apollo from reverting this data? Is there a way offix can rewrite this change back to the cache even if the request fails? It might have failed now, but it might not fail later, and just because it failed doesn't mean I no longer want to show the users the changes they have made. I don't want to lose the users changes and they should be able to sync again at a later time. I dont want to edit the cache data directly either b/c I want to allow offix to continue to do merging of local changes with new data that might get pulled down. Am I missing something?

TLDR: Offix is great, but when my request fails the optimisticResponse data is reverted by apollo. How do i fix? Thanks

Thanks,
Mitch

**Comment by darahayes** [comment]
Hey there, generally this behaviour is by design but if you want to remove that behaviour you can try removing the default queue listener. Somewhere in your application after `client.init()` has complete, you could try this:

```
delete client.queue.listeners[0].onOperationFailure
```

Let me know if that works for you!

**Comment by MitchellMonaghan** [comment]
Hi darahayes,

This didn't work for me.
1. I am using this hook to notify the user when something has gone wrong.
2. I put the code in anyway even though i need this hook and it didn't seem to have any effect.

It rollsback the data if I simply turn my browser to offline in the middle of a request. An offline app may be on a unreliable network connection (middle of no where farmers field etc) and so a some requests might succeed and others might fail. Rolling back would make the users think all their changes were just deleted but if they refresh its still there.

**Comment by MitchellMonaghan** [comment]
FWIW I am letting the user manually control when to push the local changes and not using offix to determine if they have a network connection. I'm not sure this matters tho b/c the connection could always fail in the middle of a request.

**Comment by wtrocki** [comment]
@MitchellMonaghan Amazing feedback. We have been thinking about this use case a lot. We currently using Online/Offline scheduling approach to offline where when an nonoffline error happens we have no other way to jus remove data as there is no way to send it again and we do not want to end up with something that cannot be applied to the backend. We have been exploring different use case where we going to use local database as option in the next releases - while still keeping the online-offline checks we can produce that isomorphic system that drives data from local store first. 

We will be working on POC to satisfy this case. 
In the meantime we will think how we can improve this listener in the next release to give you some capabilities that might not be related to the offix. 

The first thing I can think of is an duplication of the writes. 
Apollo GraphQL gives ability to provide local queries and writes. 
Those local queries are normalized and can be used in network queries as well. 

So to prevent from problem we can always write element to local store and ignore optimistic response features completely. This will mean that you will not need optimistic responses at all and you can use simplified API at the cost of creating local store object and removing it manually if it needs to be removed. 

Optimistic responses are typically designed for the short lived scenarios. For situations where we want data to persist despite errors we might need to use local state: https://www.apollographql.com/docs/react/data/local-state/

I would love to know what kind of error you are getting and it offix actually still manages to make request once network is restored - I kinda missed that part from your explanation

**Comment by wtrocki** [comment]
Obviously, local state will mean possible duplication of the queries for local and remote cases, which we will try to simplify itself.

**Comment by MitchellMonaghan** [comment]
My error is simply network error b/c I went offline during a request. I have a solution/workaround I wrote my self until there's better support for this scenario.

I think this might partially be my doing stuff wrong, I saw in the offix documentation there was a FetchExecutor class and I assume I should be using that instead of what I'm doing. I am overriding apollo's HttpLink fetch function as some of the api's im using are graphql and some are not (this isn't in my control).

So what I have done is I trick apollo/offix that a request was successful even though it failed b/c of no internet. I track which requests fail, as soon as one request fails I fail all other requests as the order the changes were made might be important (changed name, changed name again). When queueCleared even fires i then requeue and mutations that were not pushed to the api.
